### PR TITLE
Add support for AOF rewrite, see #147.

### PR DESCRIPTION
**Pull request by yoav-steinberg**
Still incomplete due to https://github.com/antirez/redis/issues/7350

### CODE REVIEW

**Comment by yoav-steinberg** [comment]
Note that this still isn't complete because it doesn't write the index data (aux data). See https://github.com/antirez/redis/issues/7350.
In any case we still need this because currently redis can crash if `aof-use-rdb-preamble` is set to `no` and there's no module callback implemented.

File: src/redisjson.rs [hunk 1]
```diff
@@ -503,4 +505,42 @@ pub mod type_methods {
...
+                index_name.len(),
+            );
+        } else {
+            raw::RedisModule_EmitAOF.unwrap()(
+                aof,
+                b"JSON.SET\0".as_ptr() as *const c_char,
+                b"bcb\0".as_ptr() as *const c_char,
+                key.as_ptr(),
+                key.len(),
+                b"$\0".as_ptr(),
```
Comment (gavrie) [inline src/redisjson.rs:540]: Why is the `.as_ptr()` needed, when above it's just `b"$\0"`?

File: src/redisjson.rs [hunk 2]
```diff
@@ -503,4 +505,42 @@ pub mod type_methods {
...
     }
+
+    pub unsafe extern "C" fn aof_rewrite(
+        aof: *mut raw::RedisModuleIO,
+        key: *mut raw::RedisModuleString,
+        value: *mut c_void,
+    ) {
+        let json = &*(value as *const RedisJSON);
+        let doc_str = json.data.to_string();
+        let key = RedisString::from_ptr(key).unwrap(); // FIXME: Why does it crash when I use key directly and pass "s" to the fmt string??
```
Comment (gavrie) [inline src/redisjson.rs:516]: Good question re the crash. Can you add a Redis stack race? Usually it point the problem out quite well.

File: src/redisjson.rs (hunk 1 above)
Comment (yoav-steinberg) [reply src/redisjson.rs:540]: no reason, will fix.

File: src/redisjson.rs (hunk 2 above)
Comment (yoav-steinberg) [reply src/redisjson.rs:516]: Try this on for size:
```
178714:M 13 Jul 2020 17:38:27.629 * Background append only file rewriting started by pid 178742


=== REDIS BUG REPORT START: Cut & paste starting from here ===
178742:C 13 Jul 2020 17:38:27.630 # ------------------------------------------------
178742:C 13 Jul 2020 17:38:27.630 # !!! Software Failure. Press left mouse button to continue
178742:C 13 Jul 2020 17:38:27.630 # Guru Meditation: You tried to retain an object allocated in the stack #object.c:356
178742:C 13 Jul 2020 17:38:27.630 # (forcing SIGSEGV in order to print the stack trace)
178742:C 13 Jul 2020 17:38:27.630 # ------------------------------------------------
178742:C 13 Jul 2020 17:38:27.630 # Redis 999.999.999 crashed by signal: 11
178742:C 13 Jul 2020 17:38:27.630 # Crashed running the instruction at: 0x559b0f8602e8
178742:C 13 Jul 2020 17:38:27.630 # Accessing address: 0xffffffffffffffff
178742:C 13 Jul 2020 17:38:27.630 # Failed assertion: <no assertion failed> (<no file>:0)

------ STACK TRACE ------
EIP:
redis-aof-rewrite 127.0.0.1:6379(_serverPanic+0x148)[0x559b0f8602e8]

Backtrace:
redis-aof-rewrite 127.0.0.1:6379(logStackTrace+0x4f)[0x559b0f862b4f]
redis-aof-rewrite 127.0.0.1:6379(sigsegvHandler+0xb5)[0x559b0f863355]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x153c0)[0x7fa879f793c0]
redis-aof-rewrite 127.0.0.1:6379(_serverPanic+0x148)[0x559b0f8602e8]
redis-aof-rewrite 127.0.0.1:6379(+0x5c260)[0x559b0f831260]
redis-aof-rewrite 127.0.0.1:6379(+0x5cba6)[0x559b0f831ba6]
redis-aof-rewrite 127.0.0.1:6379(moduleCreateArgvFromUserFormat+0x166)[0x559b0f895e66]
redis-aof-rewrite 127.0.0.1:6379(RM_EmitAOF+0x102)[0x559b0f896362]
target/debug/librejson.so(+0x9bb55)[0x7fa878a6ab55]
redis-aof-rewrite 127.0.0.1:6379(rewriteModuleObject+0x52)[0x559b0f85bd12]
redis-aof-rewrite 127.0.0.1:6379(rewriteAppendOnlyFileRio+0x290)[0x559b0f85c080]
redis-aof-rewrite 127.0.0.1:6379(rewriteAppendOnlyFile+0xc2)[0x559b0f85c362]
redis-aof-rewrite 127.0.0.1:6379(rewriteAppendOnlyFileBackground+0x10b)[0x559b0f85cafb]
redis-aof-rewrite 127.0.0.1:6379(startAppendOnly+0x74)[0x559b0f85ce44]
redis-aof-rewrite 127.0.0.1:6379(+0x8096a)[0x559b0f85596a]
redis-aof-rewrite 127.0.0.1:6379(+0x80cae)[0x559b0f855cae]
redis-aof-rewrite 127.0.0.1:6379(configSetCommand+0x1ba)[0x559b0f856dda]
redis-aof-rewrite 127.0.0.1:6379(configCommand+0x178)[0x559b0f8591b8]
redis-aof-rewrite 127.0.0.1:6379(call+0xb1)[0x559b0f81b311]
redis-aof-rewrite 127.0.0.1:6379(processCommand+0x34e)[0x559b0f81bace]
redis-aof-rewrite 127.0.0.1:6379(processCommandAndResetClient+0x14)[0x559b0f82a044]
redis-aof-rewrite 127.0.0.1:6379(processInputBuffer+0x18f)[0x559b0f82ea3f]
redis-aof-rewrite 127.0.0.1:6379(+0xdde5c)[0x559b0f8b2e5c]
redis-aof-rewrite 127.0.0.1:6379(aeProcessEvents+0x11a)[0x559b0f8144ea]
redis-aof-rewrite 127.0.0.1:6379(aeMain+0x2b)[0x559b0f81499b]
redis-aof-rewrite 127.0.0.1:6379(main+0x50b)[0x559b0f810f4b]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3)[0x7fa879d990b3]
redis-aof-rewrite 127.0.0.1:6379(_start+0x2e)[0x559b0f81120e]
```

Comment (yoav-steinberg) [reply src/redisjson.rs:516]: Ok, found the issue. Created a redis PR: https://github.com/redis/redis/pull/7528
Once merged I'll fix this too.

**Comment by gkorland** [comment]
not relevant anymore

**Pull request closed without merging**